### PR TITLE
LT-2372: mt: Bulk placement, editing and deleting orders

### PR DESCRIPTION
**Pull request by vashetsin**

### CODE REVIEW

File: src/MarginTrading.Backend/Controllers/OrdersController.cs [hunk 1]
```diff
@@ -35,43 +34,150 @@ namespace MarginTrading.Backend.Controllers
...
+                        Originator = request.Originator,
+                        ForceOpen = false,
+                        PositionId = position.Id,
+                        AdditionalInfo = request.AdditionalInfoJson,
+                        UseTrailingStop = request.HasTrailingStop ?? false
+                    };
+
+                    if (orderPlaceRequest.UseTrailingStop
+                        && orderPlaceRequest.Type == OrderTypeContract.StopLoss
+                        && (!string.IsNullOrWhiteSpace(orderPlaceRequest.ParentOrderId) || !string.IsNullOrWhiteSpace(orderPlaceRequest.PositionId)))
```
Comment (andrey-anokhin) [inline src/MarginTrading.Backend/Controllers/OrdersController.cs:149]: That looks strange. As I understood this action is supposed to update related orders for position only.

File: src/MarginTrading.Backend/Controllers/OrdersController.cs [hunk 2]
```diff
@@ -35,43 +34,150 @@ namespace MarginTrading.Backend.Controllers
...
+        {
+            if (!_ordersCache.Positions.TryGetPositionById(positionId, out var position))
+                throw new InvalidOperationException($"Position {positionId} not found");
+
+            var takeProfit = position.RelatedOrders?.FirstOrDefault(x => x.Type == OrderType.TakeProfit);
+            var stopLoss = position.RelatedOrders?.FirstOrDefault(x => x.Type == OrderType.StopLoss || x.Type == OrderType.TrailingStop);
+
+            var relatedOrderShouldBeRemoved = request.NewPrice == default;
+            var relatedOrderExists = request.OrderType == RelatedOrderTypeContract.TakeProfit ? takeProfit != null : stopLoss != null;
+            Func<string> relatedOrderIdFunc = () => request.OrderType == RelatedOrderTypeContract.TakeProfit
```
[the remark this answers is not in the document]
Comment (vashetsin) [reply src/MarginTrading.Backend/Controllers/OrdersController.cs:111]: when !relatedOrderShouldBeRemoved & !relatedOrderExists then calculation in advance will throw an error. to remove this function we have to duplicate in 3 different places the same logic

File: src/MarginTrading.Backend/Controllers/OrdersController.cs [hunk 3]
```diff
@@ -35,43 +34,150 @@ namespace MarginTrading.Backend.Controllers
...
+                ? takeProfit.Id
+                : stopLoss.Id;
+
+            if (!relatedOrderShouldBeRemoved)
+            {
+                if (request.OrderType == RelatedOrderTypeContract.StopLoss && relatedOrderExists)
+                {
+                    var order = _ordersCache.GetOrderById(relatedOrderIdFunc());
+                    if (order.TrailingDistance.HasValue != request.HasTrailingStop)
```
[the remark this answers is not in the document]
Comment (vashetsin) [reply src/MarginTrading.Backend/Controllers/OrdersController.cs:120]: request.OrderType has only TakeProfit & StopLoss

File: src/MarginTrading.Backend/Controllers/OrdersController.cs (hunk 1 above)
Comment (vashetsin) [reply src/MarginTrading.Backend/Controllers/OrdersController.cs:149]: this logic was copy pasted from donut

File: src/MarginTrading.Backend/Controllers/OrdersController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrey-anokhin) [reply src/MarginTrading.Backend/Controllers/OrdersController.cs:120]: I mean to check order.Type, not order.TrailingDistance.HasValue

File: src/MarginTrading.Backend/Controllers/OrdersController.cs (hunk 1 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Backend/Controllers/OrdersController.cs:149]: I understand, but we should not copy-paste wrong logic )))

**Pull request merged**